File: roboplan_rospy/roboplan_rospy/type_conversions.py
import numpy as np
from builtin_interfaces.msg import Duration
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
import roboplan


def to_duration(time_sec: float) -> Duration:
    """
    Converts a double timestamp to an equivalent ROS Duration.

    Args:
        time_sec: Timestamp in seconds.

    Returns:
        ROS 2 Duration with seconds and nanoseconds.
    """
    duration = Duration()
    duration.sec = int(time_sec)
    duration.nanosec = int((time_sec - duration.sec) * 1e9)
    return duration


def from_duration(duration: Duration) -> float:
    """
    Converts a ROS 2 Duration with seconds and nanoseconds to an equivalent double timestamp.

    Args:
        duration: ROS 2 Duration with seconds and nanoseconds.

    Returns:
        An equivalent double timestamp.
    """
    return duration.sec + duration.nanosec * 1e-9


def to_joint_trajectory(roboplan_trajectory: roboplan.JointTrajectory) -> JointTrajectory:
    """
    Converts a roboplan.JointTrajectory object to a ROS 2 JointTrajectory message.

    Args:
        roboplan_trajectory: The roboplan JointTrajectory to convert.

    Returns:
        An equivalent ROS 2 JointTrajectory message.
    """
    ros_traj = JointTrajectory()
    ros_traj.joint_names = roboplan_trajectory.joint_names

    ros_traj.points = []
    for i in range(len(roboplan_trajectory.times)):
        point = JointTrajectoryPoint()
        time_sec = roboplan_trajectory.times[i]
        point.time_from_start = to_duration(time_sec)

        if i < len(roboplan_trajectory.positions):
            point.positions = roboplan_trajectory.positions[i].tolist()

        if i < len(roboplan_trajectory.velocities):
            point.velocities = roboplan_trajectory.velocities[i].tolist()

        if i < len(roboplan_trajectory.accelerations):
            point.accelerations = roboplan_trajectory.accelerations[i].tolist()

        ros_traj.points.append(point)

    return ros_traj


def from_joint_trajectory(ros_trajectory: JointTrajectory) -> roboplan.JointTrajectory:
    """
    Convert the provided ROS 2 JointTrajectory message to an equivalent
    roboplan.JointTrajectory.

    Args:
        ros_trajectory: The ROS 2 JointTrajectory message to convert.

    Returns:
        An equivalent roboplan.JointTrajectory.
    """
    joint_traj = roboplan.JointTrajectory()

    joint_traj.joint_names = ros_trajectory.joint_names

    times = []
    positions = []
    velocities = []
    accelerations = []

    for point in ros_trajectory.points:
        times.append(from_duration(point.time_from_start))

        if point.positions:
            positions.append(np.array(point.positions))

        if point.velocities:
            velocities.append(np.array(point.velocities))

        if point.accelerations:
            accelerations.append(np.array(point.accelerations))

    joint_traj.times = times
    joint_traj.positions = positions
    joint_traj.velocities = velocities
    joint_traj.accelerations = accelerations

    return joint_traj

File: roboplan_ros_cpp/src/type_conversions.cpp
#include <roboplan_ros_cpp/type_conversions.hpp>

#include <tf2_eigen/tf2_eigen.hpp>

namespace roboplan_ros_cpp {

tl::expected<JointStateConverterMap, std::string>
buildConversionMap(const roboplan::Scene& scene, const sensor_msgs::msg::JointState& joint_state) {
  // Setup a mapping joint names to their index in the joint state
  const auto ros_joint_names = joint_state.name;
  std::unordered_map<std::string, size_t> ros_map;
  ros_map.reserve(ros_joint_names.size());
  for (size_t i = 0; i < ros_joint_names.size(); ++i) {
    ros_map[ros_joint_names[i]] = i;
  }

  // Pull out actuated joint state sizes
  const auto& model = scene.getModel();
  JointStateConverterMap conversion_map;
  conversion_map.nq = model.nq;
  conversion_map.nv = model.nv;
  conversion_map.mappings.reserve(scene.getActuatedJointNames().size());

  // For each joint name, determine type and it's location in the Scene's configuration,
  // and put it into the mapping.
  for (const auto& joint_name : scene.getJointNames()) {
    const auto joint_info = scene.getJointInfo(joint_name);
    if (!joint_info) {
      return tl::make_unexpected("Failed to get joint info for: " + joint_name);
    }
    const auto& info = joint_info.value();

    // No need to handle mimic joints
    if (info.mimic_info) {
      continue;
    }

    // TODO: Do we actually need this?
    auto it = ros_map.find(joint_name);
    if (it == ros_map.end()) {
      return tl::make_unexpected("Actuated joint: " + joint_name + ", not in list!");
    }

    // Grab the joint info and store the relevant information
    const auto joint_id = model.getJointId(joint_name);
    const auto q_idx = model.idx_qs[joint_id];
    const auto v_idx = model.idx_vs[joint_id];
    JointStateConverterMap::JointMapping mapping{.joint_name = joint_name,
                                                 .ros_index = it->second,
                                                 .q_start = static_cast<size_t>(q_idx),
                                                 .v_start = static_cast<size_t>(v_idx),
                                                 .type = info.type};
    conversion_map.mappings.push_back(mapping);
  }

  return conversion_map;
}

sensor_msgs::msg::JointState toJointState(const roboplan::JointConfiguration& joint_configuration) {
  sensor_msgs::msg::JointState joint_state;

  joint_state.name = joint_configuration.joint_names;

  joint_state.position.resize(joint_configuration.positions.size());
  for (int i = 0; i < joint_configuration.positions.size(); ++i) {
    joint_state.position[i] = joint_configuration.positions[i];
  }

  joint_state.velocity.resize(joint_configuration.velocities.size());
  for (int i = 0; i < joint_configuration.velocities.size(); ++i) {
    joint_state.velocity[i] = joint_configuration.velocities[i];
  }

  // Effort != acceleration but including and kind of abusing the interface.
  joint_state.effort.resize(joint_configuration.accelerations.size());
  for (int i = 0; i < joint_configuration.accelerations.size(); ++i) {
    joint_state.effort[i] = joint_configuration.accelerations[i];
  }

  return joint_state;
}

roboplan::JointConfiguration fromJointState(const sensor_msgs::msg::JointState& joint_state) {
  roboplan::JointConfiguration joint_configuration;

  joint_configuration.joint_names = joint_state.name;

  joint_configuration.positions =
      Eigen::VectorXd::Map(joint_state.position.data(), joint_state.position.size());

  joint_configuration.velocities =
      Eigen::VectorXd::Map(joint_state.velocity.data(), joint_state.velocity.size());

  // Effort != acceleration but including and kind of abusing the interface.
  joint_configuration.accelerations =
      Eigen::VectorXd::Map(joint_state.effort.data(), joint_state.effort.size());

  return joint_configuration;
}

trajectory_msgs::msg::JointTrajectory
toJointTrajectory(const roboplan::JointTrajectory& roboplan_trajectory) {

  trajectory_msgs::msg::JointTrajectory ros_traj;
  ros_traj.joint_names = roboplan_trajectory.joint_names;

  ros_traj.points.resize(roboplan_trajectory.times.size());
  for (size_t i = 0; i < roboplan_trajectory.times.size(); ++i) {
    auto& point = ros_traj.points[i];
    double time_sec = roboplan_trajectory.times[i];
    point.time_from_start = toDuration(time_sec);

    if (i < roboplan_trajectory.positions.size()) {
      point.positions.resize(roboplan_trajectory.positions[i].size());
      for (int j = 0; j < roboplan_trajectory.positions[i].size(); ++j) {
        point.positions[j] = roboplan_trajectory.positions[i][j];
      }
    }

    if (i < roboplan_trajectory.velocities.size()) {
      point.velocities.resize(roboplan_trajectory.velocities[i].size());
      for (int j = 0; j < roboplan_trajectory.velocities[i].size(); ++j) {
        point.velocities[j] = roboplan_trajectory.velocities[i][j];
      }
    }

    if (i < roboplan_trajectory.accelerations.size()) {
      point.accelerations.resize(roboplan_trajectory.accelerations[i].size());
      for (int j = 0; j < roboplan_trajectory.accelerations[i].size(); ++j) {
        point.accelerations[j] = roboplan_trajectory.accelerations[i][j];
      }
    }
  }

  return ros_traj;
}

roboplan::JointTrajectory
fromJointTrajectory(const trajectory_msgs::msg::JointTrajectory& ros_trajectory) {

  roboplan::JointTrajectory joint_traj;

  joint_traj.joint_names = ros_trajectory.joint_names;

  joint_traj.times.reserve(ros_trajectory.points.size());
  joint_traj.positions.reserve(ros_trajectory.points.size());
  joint_traj.velocities.reserve(ros_trajectory.points.size());
  joint_traj.accelerations.reserve(ros_trajectory.points.size());

  for (const auto& point : ros_trajectory.points) {
    joint_traj.times.push_back(fromDuration(point.time_from_start));

    Eigen::VectorXd positions =
        Eigen::VectorXd::Map(point.positions.data(), point.positions.size());
    joint_traj.positions.push_back(positions);

    Eigen::VectorXd velocities =
        Eigen::VectorXd::Map(point.velocities.data(), point.velocities.size());
    joint_traj.velocities.push_back(velocities);

    Eigen::VectorXd accelerations =
        Eigen::VectorXd::Map(point.accelerations.data(), point.accelerations.size());
    joint_traj.accelerations.push_back(accelerations);
  }

  return joint_traj;
}

geometry_msgs::msg::TransformStamped
toTransformStamped(const roboplan::CartesianConfiguration& cartesian_configuration) {
  Eigen::Isometry3d isometry(cartesian_configuration.tform);
  geometry_msgs::msg::TransformStamped ret = tf2::eigenToTransform(isometry);
  ret.header.frame_id = cartesian_configuration.base_frame;
  ret.child_frame_id = cartesian_configuration.tip_frame;
  return ret;
}

roboplan::CartesianConfiguration
fromTransformStamped(const geometry_msgs::msg::TransformStamped& transform_stamped) {
  roboplan::CartesianConfiguration config;
  config.base_frame = transform_stamped.header.frame_id;
  config.tip_frame = transform_stamped.child_frame_id;
  Eigen::Isometry3d isometry = tf2::transformToEigen(transform_stamped.transform);
  config.tform = isometry.matrix();
  return config;
}

Eigen::Matrix4d poseToSE3(const geometry_msgs::msg::Pose& pose) {
  Eigen::Isometry3d isometry;
  tf2::fromMsg(pose, isometry);
  return isometry.matrix();
}

geometry_msgs::msg::Pose se3ToPose(const Eigen::Matrix4d& transform) {
  Eigen::Isometry3d isometry(transform);
  return tf2::toMsg(isometry);
}

}  // namespace roboplan_ros_cpp
